plugins/hls-code-range-plugin/src/Ide/Plugin/CodeRange/Rules.hs

```diff
@@ -0,0 +1,186 @@
+                                                     HasStart (start))
+import           Prelude                            hiding (log)
+
+data Log = LogShake Shake.Log
+    | LogNoAST
+    | LogFoundInterleaving CodeRange CodeRange
+
+instance Pretty Log where
+    pretty log = case log of
+        LogShake shakeLog -> pretty shakeLog
+        LogNoAST          -> "no HieAst exist for file"
+        LogFoundInterleaving r1 r2 ->
+            let prettyRange = pretty . show . _codeRange_range
+             in "CodeRange interleave: " <> prettyRange r1 <> " & " <> prettyRange r2
+
+-- | A tree representing code ranges in a file. This can be useful for features like selection range and folding range
+data CodeRange = CodeRange {
+    -- | Range for current level
+        _codeRange_range    :: !Range,
+    -- | A vector of children, sorted by their ranges in ascending order.
+    -- Children are guaranteed not to interleave, but some gaps may exist among them.
+        _codeRange_children :: !(Vector CodeRange),
+    -- The kind of current code range
+        _codeRange_kind     :: !CodeRangeKind
+    }
+    deriving (Show, Generic, NFData)
+
+-- | 'CodeKind' represents the kind of a code range
+data CodeRangeKind =
+    -- | ordinary code
+    CodeKindRegion
+    -- | the group of imports
+  | CodeKindImports
+  -- | a comment
+  | CodeKindComment
+    deriving (Show, Generic, NFData)
+
+Lens.makeLenses ''CodeRange
+
+instance Eq CodeRange where
+    (==) = (==) `on` _codeRange_range
+
+instance Ord CodeRange where
+    compare :: CodeRange -> CodeRange -> Ordering
+    compare = compare `on` _codeRange_range
+
+-- | Construct a 'CodeRange'. A valid CodeRange will be returned in any case. If anything go wrong,
+-- a list of warnings will be returned as 'Log'
+buildCodeRange :: HieAST a -> RefMap a -> Annotated ParsedSource -> Writer [Log] CodeRange
+buildCodeRange ast refMap _ = do
+    -- We work on 'HieAST', then convert it to 'CodeRange', so that applications such as selection range and folding
+    -- range don't need to care about 'HieAST'
+    -- TODO @sloorush actually use 'Annotated ParsedSource' to handle structures not in 'HieAST' properly (for example comments)
+    let ast' = runReader (preProcessAST ast) (PreProcessEnv refMap)
+    codeRange <- astToCodeRange ast'
+    pure $ simplify codeRange
+
+astToCodeRange :: HieAST a -> Writer [Log] CodeRange
+astToCodeRange (Node _ sp []) = pure $ CodeRange (realSrcSpanToRange sp) mempty CodeKindRegion
+astToCodeRange node@(Node _ sp children) = do
+    children' <- removeInterleaving . sort =<< traverse astToCodeRange children
+    let codeKind = if Just CustomNodeImportsGroup == isCustomNode node then CodeKindImports else CodeKindRegion
+    pure $ CodeRange (realSrcSpanToRange sp) (V.fromList children') codeKind
+
+-- | Remove interleaving of the list of 'CodeRange's.
+removeInterleaving :: [CodeRange] -> Writer [Log] [CodeRange]
```

Comment:
I ran out of steam around here, but now this is so well factored, it might be nice to have a couple of unit tests for this function and `simplify`.

Reply:
Yes, I'm writing unit tests for every pure function like this